### PR TITLE
✨ checks/evaluation logs findings

### PR DESCRIPTION
**Pull request by laurentsimon**
closes https://github.com/ossf/scorecard/issues/3406

```release-note
Let checks/evaluation logs findings
```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -11,58 +11,183 @@ cloud.google.com/go v0.53.0/go.mod h1:fp/UouUEsRkN6ryDKNW/Upv/JBKnv6WDthjR6+vze6
 cloud.google.com/go v0.57.0/go.mod h1:oXiQ6Rzq3RAkkY7N6t3TcE6jE+CIBBbA36lwQ1JyzZs=
 cloud.google.com/go v0.110.7 h1:rJyC7nWRg2jWGZ4wSJ5nY65GTdYJkg0cd/uXb+ACI6o=
 cloud.google.com/go v0.110.7/go.mod h1:+EYjdK8e5RME/VY/qLCAtuyALQ9q67dvuum8i+H5xsI=
+cloud.google.com/go/accessapproval v1.7.1/go.mod h1:JYczztsHRMK7NTXb6Xw+dwbs/WnOJxbo/2mTI+Kgg68=
```
Comment (laurentsimon) [inline go.sum:14]: mhh, that was not intended. Will revert

Comment (laurentsimon) [reply go.sum:14]: this is strange... i've not changed this file

Comment (spencerschrock) [reply go.sum]: > this is strange... i've not changed this file

Try a `go clean -modcache`. I was having this issue recently, and it goes away at least after a clear.
https://www.github.com/golang/go/issues/61935#issuecomment-1675457180

Comment (laurentsimon) [reply go.sum]: Fixed. I just copied the one from `main` :)

**Comment by laurentsimon** [comment]
All tests green, ready for review

File: checks/evaluation/fuzzing_test.go [hunk 2]
```diff
@@ -207,7 +208,8 @@ func TestFuzzing(t *testing.T) {
 		tt := tt
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
-			if got := Fuzzing(tt.args.name, tt.args.findings); !cmp.Equal(got, tt.want, cmpopts.IgnoreFields(checker.CheckResult{}, "Error")) { //nolint:lll
+			dl := scut.TestDetailLogger{}
```
Comment (spencerschrock) [inline checks/evaluation/fuzzing_test.go:211]: should this also have TODO?

Didn't these functions previously take a DL, so we can grab the code from history?

File: checks/evaluation/security_policy.go [hunk 3]
```diff
@@ -64,9 +64,20 @@ func SecurityPolicy(name string, findings []finding.Finding) checker.CheckResult
...
 	}
 
+	if score == checker.MaxResultScore {
+		// Log all findings except the negative ones.
+		checker.LogFindings(findings, nil, []finding.Outcome{finding.OutcomeNegative}, dl)
+	} else {
+		// Log all findings.
+		checker.LogFindings(findings, nil, nil, dl)
+	}
+
```
Comment (spencerschrock) [inline checks/evaluation/security_policy.go:80]: does this check have any situations where there would be negative findings but max score?
can this be simplified to just logging findings?

File: checks/evaluation/security_policy_test.go [hunk 4]
```diff
@@ -181,7 +182,8 @@ func TestSecurityPolicy(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
 
-			got := SecurityPolicy("SecurityPolicy", tt.findings)
+			dl := scut.TestDetailLogger{}
```
Comment (spencerschrock) [inline checks/evaluation/security_policy_test.go:185]: same comment about lack of TODO on this one / being able to grab old tests?

File: checks/run_probes.go [hunk 5]
```diff
@@ -23,20 +23,15 @@ import (
 	"github.com/ossf/scorecard/v4/probes/zrunner"
 )
 
-// evaluateProbes runs the probes in probesToRun and logs its findings.
-func evaluateProbes(c *checker.CheckRequest, rawResults *checker.RawResults,
+// evaluateProbes runs the probes in probesToRun.
+func evaluateProbes(rawResults *checker.RawResults,
```
Comment (spencerschrock) [inline checks/run_probes.go:27]: without logging, all the callers could probably call zRunner directly now?

File: checker/check_result.go [hunk 6]
```diff
@@ -195,9 +197,16 @@ func CreateRuntimeErrorResult(name string, e error) CheckResult {
 }
 
 // LogFindings logs the list of findings.
-func LogFindings(findings []finding.Finding, dl DetailLogger) error {
+func LogFindings(findings []finding.Finding, findingTypesToLog, findingTypesToNotLog []finding.Outcome,
```
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checker/check_result.go:200]: So the signature could stay the same (or we could drop the `error` if we don't expect this logger to fail).
```
func LogFindings(findings []finding.Finding, dl DetailLogger) {
// same as before
}
```

And on the caller side we could do something like this (where the `nonNegative` code could live in some internal package).:

```
for i := range findings {
		f := &findings[i]
		if f.Outcome == finding.OutcomePositive {
			// Log all findings except the negative ones.
			checker.LogFindings(nonNegative(findings), dl)
			return checker.CreateMaxScoreResult(name, "update tool detected")
		}
}
```

[the remark this answers is not in the document]
Comment (spencerschrock) [reply checker/check_result.go:200]: Or alternatively, we can keep it here, but make it more generalized so we're not stuck later on by the signature which is limited to filtering on types.

What do you think of:

```
// LogFindings logs the list of findings.
func LogFindings(findings []finding.Finding, dl DetailLogger, include func(f finding.Finding) bool) error {
	for i := range findings {
		f := &findings[i]
		if !include(*f) {
			continue
		}
        // ... rest of code
```

And in some internal package we can declare some common filters:

```
func nonNegative(f finding.Finding) bool {
	return f.Outcome != finding.OutcomeNegative
}
```

[the remark this answers is not in the document]
Comment (spencerschrock) [reply checker/check_result.go:200]: I think I lead towards option 1, as it doesn't force any filtering arguments on the callers for any callers that don't want to filter.

File: checks/evaluation/dependency_update_tool_test.go [hunk 7]
```diff
@@ -211,7 +211,9 @@ func TestDependencyUpdateTool(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
 
-			got := DependencyUpdateTool(tt.name, tt.findings)
+			// TODO: verify dl results.
```
[the remark this answers is not in the document]
Comment (laurentsimon) [reply checks/evaluation/dependency_update_tool_test.go:214]: forgot to remove. dl verification is done as part of `checks/file.go`, so I don't think we need to duplicate here. Wdut?

File: checker/check_result.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (laurentsimon) [reply checker/check_result.go:200]: I prefer option 1. The current impl also works fine and is simple. I don't know what other checks will need, so I tink it's OK to keep as-is. But I can also do the Option 1 if you want. Lmk

File: checks/evaluation/fuzzing_test.go (hunk 2 above)
Comment (laurentsimon) [reply checks/evaluation/fuzzing_test.go:211]: will look into it. Good idea. Lmk about https://github.com/ossf/scorecard/pull/3409#discussion_r1300730164

File: checks/evaluation/security_policy.go (hunk 3 above)
Comment (laurentsimon) [reply checks/evaluation/security_policy.go:80]: > does this check have any situations where there would be negative findings but max score?

I don't think so.

> can this be simplified to just logging findings?

wdum?

File: checks/run_probes.go (hunk 5 above)
Comment (laurentsimon) [reply checks/run_probes.go:27]: will look into, I think you're right

File: checker/check_result.go [hunk 8]
```diff
@@ -19,6 +19,8 @@ import (
 	"fmt"
 	"math"
 
+	"golang.org/x/exp/slices"
```
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checker/check_result.go:22]: Opened #3418, so going to close this comment.

File: checker/check_result.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checker/check_result.go:200]: > That said, branch protection tiers will need the additional flexibility, so maybe worth doing now. Let's go for option 1 then?

sounds good to me

File: checks/evaluation/dependency_update_tool_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (spencerschrock) [reply checks/evaluation/dependency_update_tool_test.go:214]: It's a bit of a duplicated test yeah, but I think there's value in testing the behavior in the package where it's defined. the `checks/file.go` tests are almost like an "integration" test (but mocked).

File: checks/evaluation/security_policy.go (hunk 3 above)
Comment (spencerschrock) [reply checks/evaluation/security_policy.go:80]: > wdum?

I'm saying we can probably get rid of the if/else and conditional filtering and just have one instance of
```
checker.LogFindings(findings, dl)
```

**Comment by laurentsimon** [comment]
Ready for reviews

**Comment by spencerschrock** [comment]
> Ready for reviews

I thought we both were leaning towards option 1? https://github.com/ossf/scorecard/pull/3409#discussion_r1300711889

File: checker/check_result.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (laurentsimon) [reply checker/check_result.go]: i guess I was for option 2 then.

**Comment by laurentsimon** [comment]
> > Ready for reviews
> 
> I thought we both were leaning towards option 1? [#3409 (comment)](https://github.com/ossf/scorecard/pull/3409#discussion_r1300711889)

Sorry, I must have meant option 2.

File: checker/check_result.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (laurentsimon) [reply checker/check_result.go]: I can make the change though. I dont think I have a big pref in the end. Option 1 keep the function declaration simpler and is more flexible for the caller. So ill make this change. Sorry for last min mixup

File: checks/evaluation/dependency_update_tool_test.go [hunk 9]
```diff
@@ -211,13 +216,10 @@ func TestDependencyUpdateTool(t *testing.T) {
...
-			if tt.want.Score != got.Score {
-				t.Errorf("DependencyUpdateTool() got Score = %v, want %v for %v", got.Score, tt.want.Score, tt.name)
-			}
-			if tt.err && got.Error == nil {
-				t.Errorf("DependencyUpdateTool() error = %v, want %v for %v", got.Error, tt.want.Error, tt.name)
-				return
+			dl := scut.TestDetailLogger{}
+			got := DependencyUpdateTool(tt.name, tt.findings, &dl)
+			if !scut.ValidateTestReturn(t, tt.name, &tt.result, &got, &dl) {
+				t.Fatalf(tt.name)
```
Comment (spencerschrock) [inline checks/evaluation/dependency_update_tool_test.go:222]: change this to `t.Errorf`, one failure shouldn't fail the other tests. And the test name is already in the failure message, so can we do some form of got/want or the diff?

File: checks/evaluation/fuzzing_test.go [hunk 10]
```diff
@@ -16,199 +16,177 @@ package evaluation
...
 	for _, tt := range tests {
 		tt := tt
 		t.Run(tt.name, func(t *testing.T) {
 			t.Parallel()
-			if got := Fuzzing(tt.args.name, tt.args.findings); !cmp.Equal(got, tt.want, cmpopts.IgnoreFields(checker.CheckResult{}, "Error")) { //nolint:lll
-				t.Errorf("Fuzzing() = %v, want %v", got, cmp.Diff(got, tt.want, cmpopts.IgnoreFields(checker.CheckResult{}, "Error"))) //nolint:lll
+			dl := scut.TestDetailLogger{}
+			got := Fuzzing(tt.name, tt.findings, &dl)
+			if !scut.ValidateTestReturn(t, tt.name, &tt.result, &got, &dl) {
+				t.Fatalf(tt.name)
```
Comment (spencerschrock) [inline checks/evaluation/fuzzing_test.go:189]: same comment about not using `t.Fatalf` and tt.name is duplicate.

File: checks/evaluation/security_policy_test.go [hunk 11]
```diff
@@ -180,15 +189,10 @@ func TestSecurityPolicy(t *testing.T) {
...
-				if got.Score != -1 {
-					t.Errorf("SecurityPolicy() = %v, want %v", got, tt.want)
-				}
-			}
-			if got.Score != tt.want.Score {
-				t.Errorf("SecurityPolicy() = %v, want %v for %v", got.Score, tt.want.Score, tt.name)
+			dl := scut.TestDetailLogger{}
+			got := SecurityPolicy(tt.name, tt.findings, &dl)
+			if !scut.ValidateTestReturn(t, tt.name, &tt.result, &got, &dl) {
+				t.Fatalf(tt.name)
```
Comment (spencerschrock) [inline checks/evaluation/security_policy_test.go:195]: same comment about changing to errorf with more details

**Review by spencerschrock — APPROVED** [review]
just a few small things to take care of

**Pull request merged**